### PR TITLE
Custom Approval Conditions

### PR DESCRIPTION
**Pull request by fspmarshall**
*NOTE*: This is an **MVP** feature.  Evaluation of permissions errs on the side of strictness, and some planned features (e.g. role subselection) are not yet supported.

---

Based upon RFD/0014 (#5071)

#### Key Features

Introduces the concept of access request "reviews" which can be used to approve/deny access requests.  Unlike directly approving/denying via resource-level permissions, reviews support granular permissions and flexible multi-party approval scenarios.

Here is an example of a simple review-based configuration:

```yaml
kind: role
metadata:
  name: proletariat
spec:
  allow:
    review_requests:
      roles: ['dictator']
version: v3
---
kind: role
metadata:
  name: populist
spec:
  allow:
    request:
      roles: ['dictator']
      thresholds:
        - approve: 3
          deny: 2
version: v3
```

In the above example, users with the role `populist` are able to request the role `dictator`.  When generated, this request will automatically be configured with a review `threshold` requiring three approvals or two denials.  In this scenario, a user with role `populist` will therefore be able to assume role `dictator` if three separate users all submit approving reviews.  Users with role `proletariat` have permission to review requests for `dictator`, so three users with role `proletariat` will be able to approve a request for role `dictator` if they all agree.

Note that `thresholds` is a list.  It is possible to specify multiple thresholds.  Lets look at a more full-featured use of thresholds:

```yaml
kind: role
metadata:
  name: populist
spec:
  allow:
    request:
      roles: ['dictator']
      thresholds:
        - name: 'revolution'
          approve: 3
          deny: 2
          filter: 'contains(reviewer.roles,"proletariat")'
        - name: 'coup'
          approve: 2
          deny: 1
          filter: 'contains(reviewer.roles,"military")'
version: v3

```

In the above scenario, reviewers with the role `proletariat` count toward the `revolution` threshold, while reviewers with the role `military` count toward the `coup` threshold (reviewers with both roles count toward both thresholds).  Say, for example, that `alice` and `bob` both hold role `proletariat` and submit approving reviews.  The `revolution` threshold is not yet met, so the request is still pending.  Now say that `carol` holds role `military` and also submits an approval.  We have more approvals than either threshold requires, but neither threshold has been reached since we have one approval that matches the `coup` threshold and two approvals that match the `revolution` threshold.

*NOTE*:  A filter which matches against a reviewing user's roles (e.g. `contains(reviewer.roles, "some-role")`) effectively "leaks" the fact that the user holds that role to all other users who are allowed to view the request.  It does not, however, leak what *other* roles might be held by the reviewer, as the actual evaluation of the filter is performed only by the auth server.  The access request resource simply records a mapping of which reviews match which filters.

Reviewer permissions can also be fairly granular.  Ex:

```yaml
kind: role
metadata:
  name: proletariat
spec:
  allow:
    review_requests:
      roles: ['*']
      where: 'contains(request.system_annotations["mechanism"],"popular-uprising")'
  deny:
    review_requests:
      role: ['elected-representative']
version: v3

```

The above permits a user to review *any* access request, so long as that request does not include the role `elected-representative` and does have the system annotation `mechanism: popular-uprising`.

---

In addition to new RBAC options, `tsh` now supports a series of `request` subcommands.

Users are always allowed to see their own requests, and requests that they are allowed to review:

```
$ tsh request ls
ID                                   User  Roles    Created (UTC)       Status   
------------------------------------ ----- -------- ------------------- -------- 
094fc0f3-e5ab-47c5-816e-09326735a899 bob   dictator 30 Jan 21 03:38 UTC PENDING  
5212442f-c77c-41bb-bb14-708d2f5a0107 alice dictator 29 Jan 21 22:47 UTC APPROVED 

hint: use 'tsh request show <request-id>' for additional details
```

As indicated by the above `hint` test, `tsh request show` can be used to print out detailed information about a request:

```
$ tsh request show 5212442f-c77c-41bb-bb14-708d2f5a0107
Request ID: 5212442f-c77c-41bb-bb14-708d2f5a0107
User:       alice
Roles:      dictator
Reason:     none
Status:     APPROVED
------------------------------------------------
Approvals:
  -------------------------
  Reviewer: bob
  Reason:   "you seem cool"
  -------------------------
  Reviewer: carol
  Reason:   "very cool"
```

Reviews can be submitted via `tsh request review`:

```
$ tsh request review --deny --reason='you seem sus' 094fc0f3-e5ab-47c5-816e-09326735a899
Successfully submitted review.  Request state: DENIED
```

---

A new `access_request.review` audit event has been added:

```
{
    "event":"access_request.review",
    "id":"775ec3e0-49db-4aa6-9a4c-975763372b6f",
    "state":"PENDING",
    "reviewer":"bob",
    "proposed_state":"APPROVED"
    // ...
}
{
    "event":"access_request.review",
    "id":"775ec3e0-49db-4aa6-9a4c-975763372b6f",
    "state":"PENDING",
    "reviewer":"carol",
    "proposed_state":"APPROVED"
    // ...
}
{
    "event":"access_request.review",
    "id":"775ec3e0-49db-4aa6-9a4c-975763372b6f",
    "state":"APPROVED",
    "reviewer":"dave",
    "proposed_state":"APPROVED"
    // ...
}
```

Note that in the first two events, `state` is `PENDING` but `proposed_state` is `APPROVED`.  This indicates that the review being submitted proposes approval, but the state of the request post-application is still `PENDING`.  The third event shows state `APPROVED` indicating that the post-application state of the request is now `APPROVED`.  This is consistent with a request with an approval threshold of 3.  Subsequent reviews have no effect, but still generate events.  If, for example, the next review was a denial, we would see an event like this:

```
{
    "event":"access_request.review",
    "id":"775ec3e0-49db-4aa6-9a4c-975763372b6f",
    "state":"APPROVED",
    "reviewer":"eve",
    "proposed_state":"DENIED"
    // ...
}
```

#### Gotchas

- In order to preserve compatibility with existing configurations, `update` permissions for the `access_request` resource still allow users to approve any request.  Going forward this will not be recommended.

- Role subselection is not currently implemented for reviews.  Care has been taken to ensure that this system will be compatible with role subselection, but getting role subselection to work in a correct and intuitive manner without overly complicating the implementation is tricky.  I've opted not to let it be a blocker on the rest of the features.

- The current implementation is a little overly-strict when it comes to deciding which thresholds need to pass in order for a request to be approved, and a little over-eager to deny requests.  In particular, it requires one threshold from each relevant allow directive, and denies the entire request if any threshold from any statically assigned role is met.  In theory, this isn't always necessary, but I opted to start with the simpler and restrictive model.

- Only approval permissions from statically-assigned roles are taken into consideration.  This may become a configurable behavior in the future, but once again we're erring on the side of strictness.

#### Other Stuff

- Roles and access requests now support a `suggested_reviewers` list, and `tsh request ls` supports a `--suggested` flag which lists only requests for which the current user is a suggested reviewer.  The names in the `suggested_reviewer` list don't actually need to be teleport usernames.  They could just as easily be slack usernames, emails, etc.  This feature is purely for convenience and has no bearing on permissions.

### CODE REVIEW

**Comment by klizhentas** [comment]
@fspmarshall I thought it was a very thoughtful design, could not find anything outstanding on the first pass. I liked the suggested reviewers part as well, this is exactly what customers were asking for.

 I will do a second pass and will look in the code.

**Comment by russjones** [comment]
@awly @a-palchikov @andrejtokarcik Do you mind reviewing this PR?

File: api/types/access_request.go
Comment (klizhentas) [inline api/types/access_request.go:72]: General note to @russjones, @andrejtokarcik and @fspmarshall. This access/review/threshold system is very expressive. This raises the importance of:

https://github.com/gravitational/teleport/pull/4991 and https://github.com/gravitational/teleport/issues/3731

without tracing and role scenario testers, it will be extremely hard to troubleshoot and develop

File: tool/tsh/tsh.go
Comment (klizhentas) [inline tool/tsh/tsh.go:389]: Can you explain why do we need `no-auto-request` flag?

Comment (klizhentas) [reply tool/tsh/tsh.go:389]: Is there a way we can avoid the extra flag?

**Review by klizhentas — COMMENTED** [review]
General note to @russjones, @andrejtokarcik and @fspmarshall. This access/review/threshold system is very expressive. This raises the importance of:

#4991 and #3731

without tracing and role scenario testers, it will be extremely hard to troubleshoot and develop

**Comment by klizhentas** [comment]
@fspmarshall can you also please give an example of how to prevent access requestor from approving their own request (or if this is done automatically)? This is for the docs

**Comment by klizhentas** [comment]
Another general observation: I would explore if this code could be improved if nested loops are replaced with some auxiliary data structures, like maps, filters, or trees.

File: lib/services/access_request_test.go
Comment (a-palchikov) [inline lib/services/access_request_test.go:268]: nit: what's the reason to split these reviews into two groups for `alice`?

File: api/types/access_request.go
Comment (fspmarshall) [reply api/types/access_request.go:277]: Negative threshold IDs (indexes) don't make sense.

File: lib/services/access_request.go
Comment (fspmarshall) [reply lib/services/access_request.go:783]: Any particular reason?  We certainly don't need to support threshold counts so large that they would overflow a `uint32`.

File: lib/services/access_request.go
Comment (fspmarshall) [reply lib/services/access_request.go:548]: That field holds matchers for the request's roles, but the fact that wasn't clear definitely proves your point :)

File: tool/tsh/tsh.go
Comment (fspmarshall) [reply tool/tsh/tsh.go:389]: @klizhentas The `no-auto-request` flag is useful for debugging, but also potentially necessary as a malformed access request RBAC configuration or a misbehaving/offline plugin could otherwise lock people out of their cluster.  We could reduce clutter by making it a hidden flag, but I think its worth having.

@a-palchikov Tried your suggestion, but the result is confusing as help only shows the `auto-request` variant, which has no effect since the default is already true.  IMO this is unnecessarily confusing.

**Review by andrejtokarcik — COMMENTED** [review]
I concur with @a-palchikov's [comment on the RFD](https://github.com/gravitational/teleport/pull/5071#discussion_r541145749). We could most probably get rid of all the `RoleThresholdMapping` complexity if we conceded to not supporting actual multi-role requests. The `tsh` UI could still accept multiple specified roles but such commands would be resolved into a series of atomic requests/commands, one for each role. The issue of role subselection would immediately disappear.

This would actually facilitate one of the goals (I suppose) of introducing multi-party approvals -- to enable setups where as many requests can be reviewed by as many people as possible. With the current logic, as far as I understand, a request for roles `a,b,c` would require any potential reviewer to:
1. not be denied from reviewing *any* of `a` or `b` or `c`; and
2. be allowed to review *all* of `a` and `b` and `c`.

True, this is a result of the prudent decision to err on the side of strictness, yet it also implies that an inappropriately chosen request role could make the set of available reviewers de facto empty.

Complex role approval scenarios should be handled by external plugins that are well-understood or even developed by their users, not within Teleport itself. It's bound to be rather difficult to even document and communicate such complex solutions with all their edge cases.

Finally, I find the use of nested loops and code labels a bit too excessive here. It seems that many of the functions could be better structured by means of smaller helper functions.

**Comment by andrejtokarcik** [comment]
@klizhentas 

> can you also please give an example of how to prevent access requestor from approving their own request (or if this is done automatically)?

This looks pertinent: https://github.com/gravitational/teleport/blob/76ea3e1be64ed108dae026d647a93257ef45fe0d/lib/services/access_request.go#L385-L388

File: tool/tsh/access_request.go
Comment (andrejtokarcik) [inline tool/tsh/access_request.go:210]: Adding so much ad hoc printing code looks like an overkill. Couldn't we take advantage of `asciitable` and print a [summary table](https://github.com/gravitational/teleport/blob/09bf1ca93124e8cb3525c5c42e89cd6e8859caec/tool/tctl/common/access_request_command.go#L351-L358) of the request info followed by a table with the columns like
```
Reviewer | Decision | Submitted (UTC) | Reason
```
showing the reviews?

File: tool/tsh/tsh.go
Comment (klizhentas) [reply tool/tsh/tsh.go:389]: @fspmarshall what does `no-auto-request` actually do?

**Comment by klizhentas** [comment]
@andrejtokarcik @fspmarshall @a-palchikov re: single request to multiple request suggestion:

This would make the processing much more complex though - right now the request is a transaction with a state, it's either approved or not. With your proposal, the request would be split into multiple object, each with it's own state. This would be quite hard to reason about, as request could be partially approved/denied.

File: tool/tsh/tsh.go
Comment (fspmarshall) [reply tool/tsh/tsh.go:389]: @klizhentas It disables the automatic request generation behavior triggered by `request_access: always`.  Normally, `tsh` checks the user's roles for a `request_access` directive and automatically generates a request if the directive is present.  This is a convenience behavior and not necessary for correct functioning of `tsh`.  Given that it has the potential to lock people out of their clusters if something goes wrong with their plugins or access request configuration, I think its a behavior that people should be able to disable.

**Comment by fspmarshall** [comment]
In addition to @klizhentas's points:
 - Single-role requests would break compatibility with the plugin ecosystem.  Not just the plugins we've written, but plugins that customers have written as well.  Care has been taken to ensure that these new API elements work seamlessly with the existing plugin API s.t. they can function in tandem, or the review system can function as a fallback if plugins go offline.
 - Roles can be combined s.t. the deny directives from one role limit another.  Access requests describe a set of roles which can be assumed as a unit, which means that it is possible to have scenarios where a user isn't going to get approved for `admin` but will get approved for `admin, staging-only`.

File: api/types/access_request.go
Comment (fspmarshall) [reply api/types/access_request.go:263]: Our old model allowed changing request state after it was already approved, and that ended up being a bad decision in retrospect.  The access request system grants privileges by allowing the user to provision certificates with additional roles.  If a request is changed from approved to some other non-approved state, it gives the false impression that the privileges were never assumed.  Once a request has been in the approved state for even a few milliseconds, you cannot discount the possibility that the user has assumed elevated privileges.  Since our caching layer is fairly asynchronous, re-checking the request state on every user action isn't a great solution to this.  Its better to have the request continue to reflect the fact that the user likely did successfully escalate privilege.

In terms of cancelling reviews before the request's state-transition occurs, I'm not fundamentally opposed.  I don't see any pressing need to add it on the initial release of this feature though.

File: tool/tsh/tsh.go
Comment (klizhentas) [reply tool/tsh/tsh.go:389]: @fspmarshall what if we match 

`-request=auto|always|no`

With default value, it will figure out the reason, `--request=no` will behave similarly to agent loading behavior:

```
--add-keys-to-agent=auto|no
```

Comment (fspmarshall) [reply tool/tsh/tsh.go:389]: Per discussion with @klizhentas elsewhere, we've decided to go with using `--request-roles=no` to tell `tsh` that it should avoid creating any requests *generally* during the login sequence.  If, for some reason, a user has a role name `no` and want to supply a single-element role list with that role, they can disambiguate via a trailing comma with `--request-roles=no,`.

**Review by klizhentas — CHANGES_REQUESTED** [review]
@fspmarshall @russjones almost forgot, because it's a dual authz feature, please restrict it's usage to 

https://github.com/gravitational/teleport/blob/master/lib/modules/modules.go#L45

AdvancedAccessWorkflows

Similar to this:

https://github.com/gravitational/teleport/blob/5e9ea1c7e1ab2a9797900944edae8ea6f4b5b5cc/lib/auth/auth_with_roles.go#L1875

File: tool/tsh/access_request.go
Comment (fspmarshall) [reply tool/tsh/access_request.go:94]: Fine as-is.  While it isn't particularly *useful* to combine the currently supported filters (i.e. a combination of `--my-requests` and `--reviewable` always returns an empty list since users can't review their own requests), there is nothing inherently invalid about making that query.  Just so happens that those filters have no overlap.  Since future iterations might support filters which are useful to combine (e.g. `--created-after`), I figure its best to have filters be combinable from the beginning.

File: lib/services/access_request_test.go
Comment (fspmarshall) [reply lib/services/access_request_test.go:266]: Testing two separate scenarios is all.  Since each set of reviews are applied to the same underlying request, we need to split this scenario off into a separate block in order to start with a clean request state.

File: lib/services/access_request.go
Comment (fspmarshall) [reply lib/services/access_request.go:795]: There isn't actually any harm in double-storing (I just figure its good practice not to).  Since threshold names are displayed in the web UI, I don't want teleport to make the decision about which of two descriptions for a given requirement is the one that should be displayed.

File: api/types/access_request.go
Comment (fspmarshall) [reply api/types/access_request.go:374]: It means that this specific threshold doesn't provide a path to denial.  It doesn't prevent a different threshold from permitting denial.  This allows scenarios where some roles can only approve, while other roles can only deny.

File: api/types/role.go
Comment (fspmarshall) [reply api/types/role.go:362]: True, but 28 of the `Role` type's 33 methods make the assumption that `RoleConditionType` is a binary value (and since it is an alias of `bool`, that is currently a safe assumption).  If we decide to change that assumption, I think it is something that should be done as a separate refactor of the `Role` API.

File: lib/auth/auth_with_roles.go
Comment (fspmarshall) [reply lib/auth/auth_with_roles.go:1977]: The predicate is just a string to them.  They don't need to validate the contents of that string because they never do anything with its contents.  The auth server, on the other hand, needs to be able to evaluate the predicate.  That means that the predicate can't pull in context variables that the auth server doesn't know how pull in (e.g. if we add a new field to the review type we might start supporting `contains(request.new_field, "some-val")`).

File: lib/services/access_request_test.go
Comment (fspmarshall) [reply lib/services/access_request_test.go:97]: Thresholds without a denial count don't _prevent_ denial.  They just happen to not describe a denial condition (i.e. they don't interfere with other thresholds being capable of triggering denial).  Denial is a privilege which I prefer to have specifically granted. Consider the reverse case:

```
- name: "general disagreement"
  deny: 3
```

It would be *very* bad to silently add an `approve: 1` to that threshold.

File: tool/tsh/access_request.go
Comment (a-palchikov) [reply tool/tsh/access_request.go:94]: To me it looks as if not all filters are inherently combinable which is evident with the combination of  e.g. `--reviewable` and `--my-requests`. While `--reviewable` and `--suggested-reviews` do not exclude each other, it might still be confusing to the user as to which takes precedence. But I could be exaggerating.

File: lib/services/access_request_test.go
Comment (fspmarshall) [reply lib/services/access_request_test.go:364]: A lot of the reviews' meaningfulness comes from their position in the review list, so I ended up going with a hybrid approach.  Individual scenarios will be identified by a comment, but the reviews for that scenario will still be identified by review number.  I think this strikes a good balance between the two approaches.

File: api/types/types.proto
Comment (fspmarshall) [reply api/types/types.proto:669]: Renamed to `ProposedState`.

File: api/types/access_request.go
Comment (fspmarshall) [reply api/types/access_request.go:386]: You're not missing anything, but IMO the current layout is superior because it is deterministic regardless of the length of the review list.  If we tally all votes and then determined the state-transition as separate steps, then the result that would be produced by this logic continues to change as new reviews are added, even after the initial state-transition condition is reached.

Currently, this property is irrelevant since we skip this entire function once we've exited the `PENDING` state, but the fact that the function is deterministic means that it will be much easier to extend in the future.  Say that we wanted to add an integration with the upcoming "locking" feature which automatically generated locks if a denial theshold was triggered *after* an approval was initially triggered.  The current logic could be left basically untouched, and we could just add a new loop below that starts processing where the previous loop halted.  Basically, "its not a bug, its a feature" lol.

File: lib/auth/auth_with_roles.go
Comment (fspmarshall) [reply lib/auth/auth_with_roles.go:62]: Looked into doing this, but it seemed like the result was going to be more complexity, not less.  If we want to make `Checker`'s logging target overrideable, I think its best to tackle that as a separate issue and stick with the established pattern of using a boolean `quiet` flag for this PR.

**Review by klizhentas — APPROVED** [review]

File: lib/services/access_request.go
Comment (a-palchikov) [inline lib/services/access_request.go:502]: "not equals" comparison is suspect unless it is guaranteed that `len(approved) <= len(req.GetRoleThresholdMapping())` but it's hard to say. Would be safer to say `if len(approved) < len(req.GetRoleThresholdMapping())`.

File: lib/services/access_request.go
Comment (a-palchikov) [inline lib/services/access_request.go:727]: Accept the context as parameter instead of using `context.TODO`

File: lib/services/access_request.go
Comment (a-palchikov) [inline lib/services/access_request.go:909]:
```suggestion
		// use the defaults suggested by the user's static roles.
```

**Review by a-palchikov — APPROVED** [review]

File: lib/services/access_request.go
Comment (fspmarshall) [reply lib/services/access_request.go:335]: I agree in theory, but the difference is negligible IMO and this function is already more busy than I'd like it to be, so I think I'm gonna leave as-is for now.

File: lib/services/access_request.go
Comment (fspmarshall) [reply lib/services/access_request.go:502]: The `approved` mapping's keys are taken from the rtm, so I'm pretty confident in saying that its guaranteed.

File: api/go.mod
Comment (andrejtokarcik) [inline api/go.mod:18]: A little merge leftover.

**Review by andrejtokarcik — APPROVED** [review]

**Pull request merged**
